src/compiler/checker.ts

```diff
@@ -36802,19 +36806,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             resolutionStart = resolutionTargets.length;
         }
         links.resolvedSignature = resolvingSignature;
-        let result = resolveSignature(node, candidatesOutArray, checkMode || CheckMode.Normal);
+        const result = resolveSignature(node, candidatesOutArray, checkMode || CheckMode.Normal);
         resolutionStart = saveResolutionStart;
         // When CheckMode.SkipGenericFunctions is set we use resolvingSignature to indicate that call
         // resolution should be deferred.
         if (result !== resolvingSignature) {
-            // if the signature resolution originated on a node that itself depends on the contextual type
-            // then it's possible that the resolved signature might not be the same as the one that would be computed in source order
-            // since resolving such signature leads to resolving the potential outer signature, its arguments and thus the very same signature
-            // it's possible that this inner resolution sets the resolvedSignature first.
-            // In such a case we ignore the local result and reuse the correct one that was cached.
-            if (links.resolvedSignature !== resolvingSignature) {

```

Comment:
Is this not needed anymore? Why?

Reply:
Mostly the reason why I'm not sure is that `getResolvedSignature` ends up calling other functions besides `resolveCall`, so I'm wondering if this code can prevent circularities that don't go through `resolveCall`. But maybe not.

Reply:
I added this code as part of https://github.com/microsoft/TypeScript/pull/52146 so I already had a hunch it won't be needed when I moved this preemptive caching into `resolveCall` in this PR. In some sense, that old PR didn't move this caching early enough to fix this issue here